### PR TITLE
feat: added kyverno_policy_rule_results metric

### PR DESCRIPTION
**Pull request by yashvardhan-kukreja**
Signed-off-by: Yashvardhan Kukreja <yash.kukreja.98@gmail.com>

## Checklist

- [] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [] I have added tests that prove my fix is effective or that my feature works.
- [] I have added or changed [the documentation](https://github.com/kyverno/website).
  - If not, I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the doc update:

### CODE REVIEW

File: pkg/engine/generation.go [hunk 1]
```diff
@@ -14,17 +14,21 @@ import (
 //    - the caller has to check the ruleResponse to determine whether the path exist
 // 2. returns the list of rules that are applicable on this policy and resource, if 1 succeed
 func Generate(policyContext *PolicyContext) (resp *response.EngineResponse) {
-	return filterRules(policyContext)
+	policyStartTime := time.Now()
+	return filterRules(policyContext, policyStartTime)
 }
 
-func filterRules(policyContext *PolicyContext) *response.EngineResponse {
+func filterRules(policyContext *PolicyContext, startTime time.Time) *response.EngineResponse {
```
Comment (realshuting) [inline pkg/engine/generation.go:21]: Do we need to pass it in? Why don't we initialize startTime inside this function?

**Review by realshuting — COMMENTED** [review]
Should we log the error at the default level? Though the Prometheus errors are not critical ones, it would still be good to detect them as early as possible.

**Comment by realshuting** [comment]
What should be the best practice for the naming convention? I see we use the mixed formats of camelcase and underscore, for both package and file names. I also realized that we are not doing a good job in the kyverno repo.
 
Found this in [Effective_Go](https://golang.org/doc/effective_go#package-names):
> By convention, packages are given lower case, single-word names; there should be no need for underscores or mixedCaps.

Ideally we want to use single-word names. But if it's too long, we should at least keep them consistent.

File: pkg/engine/generation.go (hunk 1 above)
Comment (yashvardhan-kukreja) [reply pkg/engine/generation.go:21]: Yep I agree with you but semantically, it seemed more apt that the startTime(the instant when the policy jussttt starts execute) is declared exactly as the first instruction of the policy's execution which is happening at 

```go
func Generate(policyContext *PolicyContext) (resp *response.EngineResponse) {
	policyStartTime := time.Now()
	return filterRules(policyContext, policyStartTime)
}
```

**Comment by yashvardhan-kukreja** [comment]
> What should be the best practice for the naming convention? I see we use the mixed formats of camelcase and underscore, for both package and file names. I also realized that we are not doing a good job in the kyverno repo.
> 
> Found this in [Effective_Go](https://golang.org/doc/effective_go#package-names):
> 
> > By convention, packages are given lower case, single-word names; there should be no need for underscores or mixedCaps.
> 
> Ideally we want to use single-word names. But if it's too long, we should at least keep them consistent.

Good Point. You're right. It's not good to keep inconsistent naming across the project.  guess I mistook in that area. 
Looking at the project, it seems that the package names are simply written as single words without being in CamelCase or snake_case, 
And the individual Go files are written in CamelCase.

So, I believe that for this project as well we can move ahead with the same convention.

**Review by realshuting — APPROVED** [review]

**Pull request merged**